core/src/main/scala/org/apache/spark/deploy/master/Master.scala

```diff
@@ -539,7 +539,7 @@ private[deploy] class Master(
 
   private def completeRecovery() {
     // Ensure "only-once" recovery semantics using a short synchronization period.
-    if (state != RecoveryState.RECOVERING) { return }
+    if (state != RecoveryState.RECOVERING && state != RecoveryState.COMPLETING_RECOVERY) { return }
```

Comment:
@lvdongr would you please explain what will be happened after this change? With your change, which code will call this `completeRecovery`? I see there're 4 places which will call this `completeRecovery`.

Also if we have exception during recovery, should we fail fast?